File: test_app/integration_tests/ios_mocks/VoxeetSDKMock/Classes/AudioServiceAsserts.swift
import Foundation

public class AudioServiceAsserts {
    
    private static var instance: AudioServiceAsserts?
    public static func create() {
        instance = .init()
    }
    
    var methodChannel: AssertsMethodChannel?
    
    public init() {
        assertsMethodChannelFactory.createAssertMethodChannel(
            forSdkAsserts: self,
            channelName: "IntegrationTesting.AudioServiceAsserts"
        )
    }
    
    private func assertLocalStartArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.audio.local.startHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
    }
    
    private func assertLocalStopArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.audio.local.stopHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
    }

    private func assertRemoteStartArgs(args: [String: Any]) throws {
        let mockArgs = VoxeetSDK.shared.audio.remote.startArgs
        try ConferenceServiceAssertUtils.assertParticipant(args: args, mockArgs: mockArgs)
    }

    private func assertRemoteStopArgs(args: [String: Any]) throws {
        let mockArgs = VoxeetSDK.shared.audio.remote.stopArgs
        try ConferenceServiceAssertUtils.assertParticipant(args: args, mockArgs: mockArgs)
    }

    private func assertGetComfortNoiseLevelArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.audio.local.getComfortNoiseHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
        let mockArgs = VoxeetSDK.shared.audio.local.noiseLevel
        try ifKeyExists(arg: args, key: "noiseLevel") { (noiseLevel: Int) in
            try nativeAssertEquals(mockArgs.rawValue, noiseLevel)
        }
    }

    private func assertSetComfortNoiseLevelArgs(args: [String: Any]) throws {
        let mockArgs = VoxeetSDK.shared.audio.local.setComfortNoiseLevelArgs
        try ifKeyExists(arg: args, key: "noiseLevel") { (noiseLevel: Int) in
            try nativeAssertEquals(mockArgs?.rawValue, noiseLevel)
        }
    }
}

extension AudioServiceAsserts: SDKAsserts {
    
    public func runAssert(_ assert: String, args: [String: Any], result: (SDKAssertResult) -> Void) {
        do {
            switch assert {
                
            case "assertLocalStartArgs":
                try assertLocalStartArgs(args: args)
                
            case "assertLocalStopArgs":
                try assertLocalStopArgs(args: args)

            case "assertGetComfortNoiseLevelArgs":
                try assertGetComfortNoiseLevelArgs(args: args)
                
            case "assertSetComfortNoiseLevelArgs":
                try assertSetComfortNoiseLevelArgs(args: args)

            case "assertRemoteStartArgs":
                try assertRemoteStartArgs(args: args)

            case "assertRemoteStopArgs":
                try assertRemoteStopArgs(args: args)

            default:
                throw SDKAssertError.unknownAssert
            }
            result(.success(()))
        } catch {
            result(SDKAssertError.toResult(error: error))
        }
    }
}

File: test_app/integration_tests/android_mocks/src/main/java/com/voxeet/asserts/AudioServiceAsserts.java
package com.voxeet.asserts;

import com.voxeet.VoxeetSDK;
import com.voxeet.android.media.utils.ComfortNoiseLevel;
import com.voxeet.sdk.models.Participant;

import java.util.Map;

import io.dolby.asserts.AssertUtils;
import io.dolby.asserts.MethodDelegate;

public class AudioServiceAsserts implements MethodDelegate {

    @Override
    public void onAction(String methodName, Map<String, Object> args, Result result) {
        try {
            switch (methodName) {
                case "assertLocalStartArgs":
                    assertLocalStartArgs(args);
                    break;
                case "assertLocalStopArgs":
                    assertLocalStopArgs(args);
                    break;
                case "assertRemoteStartArgs":
                    assertRemoteStartArgs(args);
                    break;
                case "assertRemoteStopArgs":
                    assertRemoteStopArgs(args);
                    break;
                case "assertGetComfortNoiseLevelArgs":
                    assertGetComfortNoiseLevelArgs(args);
                    break;
                case "assertSetComfortNoiseLevelArgs":
                    assertSetComfortNoiseLevelArgs(args);
                    break;
                default:
                    result.error(new NoSuchMethodError("Method: " + methodName + " not found in " + getName() + " method channel"));
                    return;
            }
            result.success();
        } catch (AssertionFailed exception) {
            result.failed(exception);
        } catch (Exception ex) {
            result.error(ex);
        }
    }

    @Override
    public String getName() {
        return "IntegrationTesting.AudioServiceAsserts";
    }

    private void assertLocalStartArgs(Map<String, Object> args) throws KeyNotFoundException, MethodDelegate.AssertionFailed {
        boolean mockHasRun = VoxeetSDK.audio().local.startHasRun;
        if (args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "audio local not started");
        } else {
            throw new KeyNotFoundException("hasRun key not found");
        }
    }

    private void assertLocalStopArgs(Map<String, Object> args) throws KeyNotFoundException, MethodDelegate.AssertionFailed {
        boolean mockHasRun = VoxeetSDK.audio().local.stopHasRun;
        if (args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "audio local not stopped");
        } else {
            throw new KeyNotFoundException("hasRun key not found");
        }
    }

    private void assertRemoteStartArgs(Map<String, Object> args) throws KeyNotFoundException, MethodDelegate.AssertionFailed {
        Participant mockHasRun = VoxeetSDK.audio().remote.startArgs;
        AssertionHelper.assertParticipant(args, mockHasRun);
    }

    private void assertRemoteStopArgs(Map<String, Object> args) throws KeyNotFoundException, MethodDelegate.AssertionFailed {
        Participant mockHasRun = VoxeetSDK.audio().remote.stopArgs;
        AssertionHelper.assertParticipant(args, mockHasRun);
    }

    private void assertGetComfortNoiseLevelArgs(Map<String, Object> args) throws KeyNotFoundException, MethodDelegate.AssertionFailed {
        boolean mockHasRun = VoxeetSDK.mediaDevice().getComfortNoiseLevelHasRun;
        if (args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "getComfortNoise not run");
        }
        int mockArgs = VoxeetSDK.mediaDevice().level.value();
        if (args.containsKey("noiseLevel")) {
            AssertUtils.compareWithExpectedValue(mockArgs, (Integer) args.get("noiseLevel"), "incorrect noise level");
        }
    }

    private void assertSetComfortNoiseLevelArgs(Map<String, Object> args) throws KeyNotFoundException, MethodDelegate.AssertionFailed {
        int mockArgs = VoxeetSDK.mediaDevice().level.value();
        if (args.containsKey("noiseLevel")) {
            AssertUtils.compareWithExpectedValue(mockArgs, (Integer) args.get("noiseLevel"), "incorrect noise level");
        }
    }
}
